### PR TITLE
FK progress, INSERT referencing works

### PR DESCRIPTION
**Pull request by Hydrocharged**
Referencing works, so `INSERT INTO child VALUES (1, 2, 1);` will check the `parent` table for the appropriate columns using an index. Cascading will be handled similarly, with each `UPDATE` or `DELETE` passing a row to a child that then recursively updates or deletes along the tree. MySQL uses DFS for cascading so that should work.

I wouldn't worry about the failing tests, a lot of existing tests are wrong, but haven't been caught until now due to harness and implementation differences in GMS and Dolt. I'll update them later.

### CODE REVIEW

File: sql/plan/foreign_key_check.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+//TODO: doc and check field names
+type ForeignKeyReferenceChecker struct {
```
Comment (andrew-wm-arthur) [inline sql/plan/foreign_key_check.go:26]: Is this node going to be responsible to checking parent references and cascading to children? or will those be separate nodes?

**Review by andrew-wm-arthur — COMMENTED** [review]
The general direction here makes sense. Are you at a point where we can feature flag this work and get it merge into main?

File: sql/plan/foreign_key_check.go (hunk 1 above)
Comment (Hydrocharged) [reply sql/plan/foreign_key_check.go:26]: They’re going to be separate nodes at first, and then if I see a large common path between the two I’ll merge them, I think that’s the easiest developmental approach.

**Comment by Hydrocharged** [comment]
Changes were manually added into the base PR.

**Pull request closed without merging**